client/fallback_provider.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2022 Lekko Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"errors"
+
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+func FallbackProvider(primary Provider, backup Provider) Provider {
+	return &fallbackProvider{primary, backup}
+}
+
+type fallbackProvider struct {
+	primary Provider
+	backup  Provider
+}
+
+func (p *fallbackProvider) Close(ctx context.Context) error {
+	primaryErr := p.primary.Close(ctx)
+	backupErr := p.backup.Close(ctx)
+	if primaryErr != nil && backupErr != nil {
```

Comment:
I think that you can just join them, `errors.Join` discards `nil`: https://pkg.go.dev/errors#Join